### PR TITLE
Refactor/delete me after fix

### PR DESCRIPTION
**Pull request by StegSchreck**

### CODE REVIEW

**Comment by StegSchreck** [comment]
@Sh4kE shall we merge this to `refactor/add_number`?

File: src/main/kotlin/sudoklin/data/Sudoku.kt [hunk 1]
```diff
@@ -112,10 +112,13 @@ class SudokuPuzzle (val matrix: Array<Array<IntArray>>) {
...
-class NewShinySudoku() {
-    val rows: Array<SudokuRow?> = Array(9, {_ -> null})
-    val columns: Array<SudokuColumn?> = Array(9, {_ -> null})
-    val groups: Array<SudokuGroup?> = Array(9, {_ -> null})
+class NewShinySudoku(val rows: Array<SudokuRow?> = Array(9, {_ -> null}),
+                     val columns: Array<SudokuColumn?> = Array(9, {_ -> null}),
+                     val groups: Array<SudokuGroup?> = Array(9, {_ -> null})) {
+
+    fun clone(): NewShinySudoku {
+        return NewShinySudoku(rows, columns, groups)
```
Comment (StegSchreck) [inline src/main/kotlin/sudoklin/data/Sudoku.kt]: are the arrays cloned, too?

File: src/main/kotlin/sudoklin/data/Sudoku.kt [hunk 2]
```diff
@@ -147,6 +150,29 @@ class NewShinySudoku() {
...
+            }
+        }
+        return true
+    }
+
+    fun isValid(): Boolean {
+        var isValid = true
+        for (i in 0..8) {
+            if (!rows[i]!!.isValid())
+                isValid = false
```
Comment (StegSchreck) [inline src/main/kotlin/sudoklin/data/Sudoku.kt]: we could directly `return false` here

File: src/main/kotlin/sudoklin/solver/SudokuSolver.kt [hunk 3]
```diff
@@ -183,4 +184,181 @@ class SudokuSolver(initialSudoku: Sudoku, val log: Boolean = false) {
             }
         }
     }
+}
+
+class NewSudokuSolver(val sudoku: NewShinySudoku, val log: Boolean = false) {
+    val timestamp: String = (System.currentTimeMillis() / (60 * 1000)).toString()
+
+    fun getMissingNumbersForRow(rowIndex: Int): List<Int> {
```
Comment (StegSchreck) [inline src/main/kotlin/sudoklin/solver/SudokuSolver.kt]: can we merge `getMissingNumbersForRow`, `getMissingNumbersForColumn` and `getMissingNumbersForGroup` to reduce duplicated code?

File: src/main/kotlin/sudoklin/solver/SudokuSolver.kt [hunk 4]
```diff
@@ -183,4 +184,181 @@ class SudokuSolver(initialSudoku: Sudoku, val log: Boolean = false) {
...
+        }
+        return sudoku
+    }
+
+    private fun fillNumbersToTheirOnlyPossibleCell() {
+        for (number in 1..9) {
+            for (i in 0..8) {
+                val possiblePositionsInRow = getPossiblePositionsForNumberInRow(number, i)
+                if (possiblePositionsInRow.size == 1) {
+                    sudoku.rows[i]!!.cells[possiblePositionsInRow[0]].candidates.clear()
```
Comment (StegSchreck) [inline src/main/kotlin/sudoklin/solver/SudokuSolver.kt]: we are always doing `clear` and then `add`. Let's have one convenience function which wraps this.

File: src/main/kotlin/sudoklin/solver/SudokuSolver.kt [hunk 5]
```diff
@@ -220,67 +220,30 @@ class NewSudokuSolver(val sudoku: NewShinySudoku, val log: Boolean = false) {
...
-
-                    val groupIndex = sudoku.getCell(rowIndex, columnIndex).groupIndex
-                    val alreadyInGoup = sudoku.groups[groupIndex]!!.containsSolvedNumber(number)
-
-                    if (! (alreadyInColumn || alreadyInGoup)) {
-                        possiblePositions.add(columnIndex)
-                    }
-                }
-            }
+            return sudoku.rows[rowIndex]!!.cells.filter { it.candidates.contains(number) }.map { it.columnIndex }
```
Comment (StegSchreck) [inline src/main/kotlin/sudoklin/solver/SudokuSolver.kt]: nice!

File: src/main/kotlin/sudoklin/solver/SudokuSolver.kt [hunk 6]
```diff
@@ -324,7 +287,10 @@ class NewSudokuSolver(val sudoku: NewShinySudoku, val log: Boolean = false) {
 //            /*1*/ fillOnlyPossibleValueForCells()
 //            if (log) File("./$timestamp.log").appendText("Step $iteration a:" + workingStack.last().puzzle.toString())
 
-            /*2*/ fillNumbersToTheirOnlyPossibleCell()
+            /*2*/
+            val changed = fillNumbersToTheirOnlyPossibleCell()
+            if (!changed)
+                return sudoku
```
Comment (StegSchreck) [inline src/main/kotlin/sudoklin/solver/SudokuSolver.kt]: ?

File: src/test/kotlin/sudoklin/solver/NewSudokuSolverTest.kt [hunk 7]
```diff
@@ -133,15 +132,15 @@ class NewSudokuSolverTest : Spek({
                     assertTrue(solvedSudoku.isValid())
                 }
 
-/*                it("should be able to solve an medium puzzle") {
+                it("should be able to solve an medium puzzle") {
```
Comment (StegSchreck) [inline src/test/kotlin/sudoklin/solver/NewSudokuSolverTest.kt]: this test is failing.
Do we need more advanced algorithms before we can solve this?

File: src/main/kotlin/sudoklin/solver/SudokuSolver.kt [hunk 8]
```diff
@@ -318,4 +316,32 @@ class NewSudokuSolver(val sudoku: NewShinySudoku, val log: Boolean = false) {
...
+            changed = findPairsOfCandidatesInList(list)
+        }
+        return changed
+    }
+
+    private fun findPairsOfCandidatesInList(list: SudokuList?): Boolean {
+        var changed = false
+        val twoItemCellsInList = list!!.cells.filter { it.candidates.size == 2 }
+        for (cell in twoItemCellsInList) {
+            val samezies = twoItemCellsInList.filterNot { it == cell }.filter { it.candidates.containsAll(cell.candidates) }
```
Comment (Sh4kE) [inline src/main/kotlin/sudoklin/solver/SudokuSolver.kt]: How would you pronounce samezies? ;)

**Pull request merged**